tests/test_io_failure.py

```diff
@@ -0,0 +1,121 @@
+"""
+
+
+def generate_repo(repo_path, akt_repo):
+    arepo = [akt_repo, '--keytype', 'dummy']
+    run([*arepo, 'generate', '--path', repo_path])
+    fw_path = path.join(repo_path, 'images/firmware.txt')
+    os.makedirs(path.join(repo_path, 'images'))
+    with open(fw_path, 'wb') as f:
+        f.write(b'fw')
+    run([*arepo, 'image', '--path', repo_path, '--filename', fw_path,
+         '--targetname', 'firmware.txt'])
+    run([*arepo, 'addtarget', '--path', repo_path, '--targetname',
+         'firmware.txt', '--hwid', 'primary_hw', '--serial', 'CA:FE:A6:D2:84:9D'])
+    run([*arepo, 'signtargets', '--path', repo_path])
+
+
+def run_test(akt_test, server, srcdir, pf):
+    with tempfile.TemporaryDirectory() as storage_dir:
+        # run once with (maybe) a fault, trying to update
+        cp_err = run(['fiu-run', '-x', '-c', f'enable_random name=posix/io/*,probability={pf}',
+                      akt_test, storage_dir, server],
+                     stdout=subprocess.PIPE, stderr=subprocess.STDOUT, cwd=srcdir)
+        if cp_err.returncode == 0:
+            return True
+        print('update failed, checking state')
+        cp = run([akt_test, storage_dir, server],
+                 stdout=subprocess.PIPE, stderr=subprocess.STDOUT, cwd=srcdir)
+        if cp.returncode != 0:
+            error_state_dir = f'fail_state.{path.basename(storage_dir)}'
+            print(f'Error detected, see {error_state_dir}')
+            shutil.copytree(storage_dir, error_state_dir)
+            with open(path.join(error_state_dir, 'output1.log'), 'wb') as f:
+                f.write(cp_err.stdout)
+            with open(path.join(error_state_dir, 'output2.log'), 'wb') as f:
+                f.write(cp.stdout)
+            return False
+        return True
+
+
+def pooled_test(f, k):
+    print(f'Running test {k}')
+    return f()
+
+
+def main():
+    parser = argparse.ArgumentParser(description='Run io failure tests')
+    parser.add_argument('-n', '--n-tests', type=int, default=100,
+                        help='number of random tests to run')
+    parser.add_argument('-j', '--jobs', type=int, default=1,
+                        help='number of parallel tests')
+    parser.add_argument('-pf', '--probability-failure', type=float, default=0.00025,
+                        help='probability of syscall failure')
+    parser.add_argument('--akt-srcdir', help='path to the aktualizr source directory')
+    parser.add_argument('--akt-repo', help='path to aktualizr-repo executable')
+    parser.add_argument('--akt-test', help='path to aktualizr cycle test')
+    parser.add_argument('--serve-only', action='store_true',
+                        help='only serve metadata, do not run tests')
+    args = parser.parse_args()
+
+    srcdir = path.abspath(args.akt_srcdir) if args.akt_srcdir is not None else os.getcwd()
+
+    print(f'srcdir {srcdir}')
+
+    with tempfile.TemporaryDirectory() as repo_dir:
```

Comment:
Why the temp dir and repo generation can't be hidden/done within FakeTestServerBackground ?, it's not important for a test, though.

Reply:
Keeping it loosely coupled seemed more straightforward and easier to refactor/reuse.